excluded variants break build

The plugin adds dependencies for tasks of build types which are excluded and therefore not found.

```gradle
android {
    ...
    variantFilter { variant ->
        if (variant.name == "debug") {
            variant.setIgnore(true)
        }
    }
}
```

Results in:

```
FAILURE: Build failed with an exception.

* What went wrong:
Could not determine the dependencies of task ':app:jacocoTestReportDebug'.
> Task with path 'testDebugUnitTest' not found in project ':app'.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

BUILD FAILED
```

--- a/src/main/groovy/com/vanniktech/android/junit/jacoco/Generation.groovy
+++ b/src/main/groovy/com/vanniktech/android/junit/jacoco/Generation.groovy
@@ -1,5 +1,6 @@
 package com.vanniktech.android.junit.jacoco
 
+import com.android.build.gradle.api.BaseVariant
 import org.gradle.api.Plugin
 import org.gradle.api.Project
 import org.gradle.testing.jacoco.tasks.JacocoReport
@@ -28,9 +29,10 @@ class Generation implements Plugin<Project> {
         }
     }
 
-    protected static boolean addJacoco(final Project subProject, final JunitJacocoExtension extension) {
+    protected static boolean addJacoco(final Project subProject,
+            final JunitJacocoExtension extension) {
         if (!shouldIgnore(subProject, extension)) {
-            if (isAndroidProject(subProject)) {
+            if (isAndroidApplication(subProject) || isAndroidLibrary(subProject)) {
                 addJacocoAndroid(subProject, extension)
                 return true
             } else if (isJavaProject(subProject)) {
@@ -84,65 +86,69 @@ class Generation implements Plugin<Project> {
             toolVersion extension.jacocoVersion
         }
 
-        subProject.android.testOptions.unitTests.all {
+        subProject.android.testOptions?.unitTests?.all {
             it.jacoco.includeNoLocationClasses = extension.includeNoLocationClasses
         }
 
-        final def buildTypes = subProject.android.buildTypes.collect { type -> type.name }
-        final def productFlavors = subProject.android.productFlavors.collect { flavor -> flavor.name }
+        Collection<BaseVariant> variants = []
+        if (isAndroidApplication(subProject)) {
+            variants = subProject.android.applicationVariants
+        } else if (isAndroidLibrary(subProject)) {
+            variants = subProject.android.libraryVariants
+        }
 
-        // When no product flavors defined, use empty
-        if (!productFlavors) productFlavors.add('')
+        variants.all { variant ->
 
-        productFlavors.each { productFlavorName ->
-            buildTypes.each { buildTypeName ->
+            def productFlavorName = variant.getFlavorName()
+            def buildTypeName = variant.getBuildType().name
 
-                def sourceName, sourcePath
-                if (!productFlavorName) {
-                    sourceName = sourcePath = "${buildTypeName}"
-                } else {
-                    sourceName = "${productFlavorName}${buildTypeName.capitalize()}"
-                    sourcePath = "${productFlavorName}/${buildTypeName}"
-                }
-                final def testTaskName = "test${sourceName.capitalize()}UnitTest"
-                final def taskName = "jacocoTestReport${sourceName.capitalize()}"
-
-                subProject.task(taskName, type: JacocoReport, dependsOn: testTaskName) {
-                    group = 'Reporting'
-                    description = "Generate Jacoco coverage reports after running ${sourceName} tests."
-
-                    reports {
-                        xml {
-                            enabled = true
-                            destination "${subProject.buildDir}/reports/jacoco/${sourceName}/jacoco.xml"
-                        }
-                        html {
-                            enabled = true
-                            destination "${subProject.buildDir}/reports/jacoco/${sourceName}"
-                        }
-                    }
-
-                    classDirectories = subProject.fileTree(
-                            dir: "${subProject.buildDir}/intermediates/classes/${sourcePath}",
-                            excludes: getExcludes(extension)
-                    )
-
-                    final def coverageSourceDirs = [
-                            "src/main/java",
-                            "src/$buildTypeName/java"
-                    ]
-
-                    if (productFlavorName) {
-                        coverageSourceDirs.add("src/$productFlavorName/java")
-                    }
-
-                    additionalSourceDirs = subProject.files(coverageSourceDirs)
-                    sourceDirectories = subProject.files(coverageSourceDirs)
-                    executionData = subProject.files("${subProject.buildDir}/jacoco/${testTaskName}.exec")
-                }
-
-                subProject.check.dependsOn "${taskName}"
+            def sourceName, sourcePath
+            if (!productFlavorName) {
+                sourceName = sourcePath = "${buildTypeName}"
+            } else {
+                sourceName = "${productFlavorName}${buildTypeName.capitalize()}"
+                sourcePath = "${productFlavorName}/${buildTypeName}"
             }
+
+            final def testTaskName = "test${sourceName.capitalize()}UnitTest"
+            final def taskName = "jacocoTestReport${sourceName.capitalize()}"
+
+            subProject.task(taskName, type: JacocoReport, dependsOn: testTaskName) {
+                group = 'Reporting'
+                description = "Generate Jacoco coverage reports after running ${sourceName} tests."
+
+                reports {
+                    xml {
+                        enabled = true
+                        destination "${subProject.buildDir}/reports/jacoco/${sourceName}/jacoco.xml"
+                    }
+                    html {
+                        enabled = true
+                        destination "${subProject.buildDir}/reports/jacoco/${sourceName}"
+                    }
+                }
+
+                classDirectories = subProject.fileTree(
+                        dir: "${subProject.buildDir}/intermediates/classes/${sourcePath}",
+                        excludes: getExcludes(extension)
+                )
+
+                final def coverageSourceDirs = [
+                        "src/main/java",
+                        "src/$buildTypeName/java"
+                ]
+
+                if (productFlavorName) {
+                    coverageSourceDirs.add("src/$productFlavorName/java")
+                }
+
+                additionalSourceDirs = subProject.files(coverageSourceDirs)
+                sourceDirectories = subProject.files(coverageSourceDirs)
+                executionData =
+                        subProject.files("${subProject.buildDir}/jacoco/${testTaskName}.exec")
+            }
+
+            subProject.check.dependsOn "${taskName}"
         }
     }
 
@@ -169,10 +175,12 @@ class Generation implements Plugin<Project> {
         ] : extension.excludes
     }
 
-    protected static boolean isAndroidProject(final Project project) {
-        final boolean isAndroidLibrary = project.plugins.hasPlugin('com.android.library')
-        final boolean isAndroidApp = project.plugins.hasPlugin('com.android.application')
-        return isAndroidLibrary || isAndroidApp
+    protected static boolean isAndroidLibrary(final Project project) {
+        return project.plugins.hasPlugin('com.android.library')
+    }
+
+    protected static boolean isAndroidApplication(final Project project) {
+        return project.plugins.hasPlugin('com.android.application')
     }
 
     protected static boolean isJavaProject(final Project project) {

--- a/src/main/groovy/com/vanniktech/android/junit/jacoco/Generation.groovy
+++ b/src/main/groovy/com/vanniktech/android/junit/jacoco/Generation.groovy
@@ -144,8 +144,7 @@ class Generation implements Plugin<Project> {
 
                 additionalSourceDirs = subProject.files(coverageSourceDirs)
                 sourceDirectories = subProject.files(coverageSourceDirs)
-                executionData =
-                        subProject.files("${subProject.buildDir}/jacoco/${testTaskName}.exec")
+                executionData = subProject.files("${subProject.buildDir}/jacoco/${testTaskName}.exec")
             }
 
             subProject.check.dependsOn "${taskName}"

--- a/src/test/groovy/com/vanniktech/android/junit/jacoco/ProjectHelper.groovy
+++ b/src/test/groovy/com/vanniktech/android/junit/jacoco/ProjectHelper.groovy
@@ -1,5 +1,10 @@
 package com.vanniktech.android.junit.jacoco
 
+import com.android.build.gradle.AppExtension
+import com.android.build.gradle.LibraryExtension
+import com.android.build.gradle.api.BaseVariant
+import com.android.builder.model.BuildType
+import groovy.mock.interceptor.MockFor
 import org.gradle.api.Project
 import org.gradle.testfixtures.ProjectBuilder
 
@@ -21,9 +26,45 @@ final class ProjectHelper {
                 break
             case ProjectType.ANDROID_APPLICATION:
                 project = ProjectBuilder.builder().withName('android app').build()
+                def androidMock = new MockFor(AppExtension)
+                def buildTypesMock = ["debug", "release"].collect { bt ->
+                    def type = new MockFor(BuildType)
+                    type.metaClass.getName = { bt }
+                    type
+                }
+                androidMock.metaClass.getBuildTypes = { buildTypesMock }
+                def appVariants = buildTypesMock.collect { bt ->
+                    def variant = new MockFor(BaseVariant)
+                    variant.metaClass.getFlavorName = { null }
+                    variant.metaClass.getBuildType = { bt }
+                    variant
+                }
+                androidMock.metaClass.getApplicationVariants = { appVariants }
+                androidMock.metaClass.testOptions = null
+                project.metaClass.android = androidMock
+                // fake all with each
+                project.android.applicationVariants.metaClass.all = { delegate.each(it) }
                 break
             case ProjectType.ANDROID_LIBRARY:
                 project = ProjectBuilder.builder().withName('android library').build()
+                def androidMock = new MockFor(LibraryExtension)
+                def buildTypesMock = ["debug", "release"].collect { bt ->
+                    def type = new MockFor(BuildType)
+                    type.metaClass.getName = { bt }
+                    type
+                }
+                androidMock.metaClass.getBuildTypes = { buildTypesMock }
+                def appVariants = buildTypesMock.collect { bt ->
+                    def variant = new MockFor(BaseVariant)
+                    variant.metaClass.getFlavorName = { null }
+                    variant.metaClass.getBuildType = { bt }
+                    variant
+                }
+                androidMock.metaClass.getLibraryVariants = { appVariants }
+                androidMock.metaClass.testOptions = null
+                project.metaClass.android = androidMock
+                // fake all with each
+                project.android.libraryVariants.metaClass.all = { delegate.each(it) }
                 break
         }
 
@@ -41,12 +82,29 @@ final class ProjectHelper {
                 blue: [applicationId: 'com.example.blue']
         ]
 
-        project.android.productFlavors {
-            customFlavors.each { name, config ->
-                "$name" {
-                    applicationId config.applicationId
+        def variants = customFlavors.collect { flavorName, config ->
+            project.android.buildTypes.collect { buildType ->
+                def variant = new MockFor(BaseVariant)
+                variant.metaClass.getBuildType = {
+                    def type = new MockFor(BuildType)
+                    type.metaClass.getName = { buildType.name }
+                    type
                 }
+                variant.metaClass.getFlavorName = { flavorName }
+                variant.metaClass.getApplicationId = { config.applicationId }
+                variant
             }
+        }.flatten()
+
+        switch (projectType) {
+            case ProjectType.ANDROID_APPLICATION:
+                project.android.metaClass.applicationVariants = variants
+                project.android.applicationVariants.metaClass.all = { delegate.each(it) }
+                break
+            case ProjectType.ANDROID_LIBRARY:
+                project.android.metaClass.libraryVariants = variants
+                project.android.libraryVariants.metaClass.all = { delegate.each(it) }
+                break
         }
 
         return this

--- a/src/test/groovy/com/vanniktech/android/junit/jacoco/ProjectHelper.groovy
+++ b/src/test/groovy/com/vanniktech/android/junit/jacoco/ProjectHelper.groovy
@@ -42,7 +42,7 @@ final class ProjectHelper {
                 androidMock.metaClass.getApplicationVariants = { appVariants }
                 androidMock.metaClass.testOptions = null
                 project.metaClass.android = androidMock
-                // fake all with each
+                // mock .all{ } function from android gradle lib with standard groovy .each{ }
                 project.android.applicationVariants.metaClass.all = { delegate.each(it) }
                 break
             case ProjectType.ANDROID_LIBRARY:
@@ -63,7 +63,7 @@ final class ProjectHelper {
                 androidMock.metaClass.getLibraryVariants = { appVariants }
                 androidMock.metaClass.testOptions = null
                 project.metaClass.android = androidMock
-                // fake all with each
+                // mock .all{ } function from android gradle lib with standard groovy .each{ }
                 project.android.libraryVariants.metaClass.all = { delegate.each(it) }
                 break
         }
@@ -99,10 +99,12 @@ final class ProjectHelper {
         switch (projectType) {
             case ProjectType.ANDROID_APPLICATION:
                 project.android.metaClass.applicationVariants = variants
+                // mock .all{ } function from android gradle lib with standard groovy .each{ }
                 project.android.applicationVariants.metaClass.all = { delegate.each(it) }
                 break
             case ProjectType.ANDROID_LIBRARY:
                 project.android.metaClass.libraryVariants = variants
+                // mock .all{ } function from android gradle lib with standard groovy .each{ }
                 project.android.libraryVariants.metaClass.all = { delegate.each(it) }
                 break
         }

